rust/src/treefile.rs

```diff
@@ -0,0 +1,191 @@
+ * */
+
+extern crate libc;
+
+#[macro_use]
+extern crate serde_derive;
+extern crate serde;
+extern crate serde_json;
+extern crate serde_yaml;
+extern crate tempfile;
+
+use std::fs;
+use std::io;
+use std::io::Write;
+use std::ffi::CStr;
+use std::ffi::OsStr;
+use std::path::Path;
+use std::os::unix::ffi::OsStrExt;
+use std::os::unix::io::FromRawFd;
+
+#[no_mangle]
+pub extern fn treefile_read(filename: *const libc::c_char, fd: libc::c_int) -> libc::c_int {
+
+    // using an O_TMPFILE is an easy way to avoid ownership transfer issues w/ returning allocated
+    // memory across the Rust/C boundary
+
+    // let's just get the unsafory stuff (see what I did there?) out of the way first
+    let output_file = unsafe { fs::File::from_raw_fd(libc::dup(fd)) };
+    let c_str: &CStr = unsafe { CStr::from_ptr(filename) };
+    let filename_path = Path::new(OsStr::from_bytes(c_str.to_bytes()));
+
+    match treefile_read_impl(filename_path, output_file) {
+        Err(e) => {
+            // just write it out to stderr for now
+            let _ = writeln!(io::stderr(), "treefile error: {}", e);
+            0
+        },
+        _ => 1
+    }
+}
+
+fn treefile_read_impl(filename: &Path, output: fs::File) -> io::Result<()> {
+    let f = fs::File::open(filename)?;
+
+    let mut treefile: TreeComposeConfig = match serde_yaml::from_reader(f) {
+        Ok(t) => t,
+        Err(e) => return Err(io::Error::new(io::ErrorKind::InvalidInput, format!("{}", e))),
+    };
+
+    /* special handling for packages, since we allow whitespaces within items */
+    if let Some(pkgs) = treefile.packages {
+        treefile.packages = Some(whitespace_split_packages(&pkgs));
+    }
+
+    serde_json::to_writer_pretty(output, &treefile)?;
+
+    Ok(())
+}
+
+fn whitespace_split_packages(pkgs: &Vec<String>) -> Vec<String> {
+    let mut ret = Vec::with_capacity(pkgs.len());
+    for pkg in pkgs {
+        for pkg_item in pkg.split_whitespace() {
+            ret.push(pkg_item.into());
+        }
+    }
+    return ret;
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum BootLocation {
+    #[serde(rename = "both")]
+    Both,
+    #[serde(rename = "legacy")]
+    Legacy,
+    #[serde(rename = "new")]
+    New,
+}
+
+impl Default for BootLocation {
+    fn default() -> Self {
+        BootLocation::Both
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum CheckPasswdType {
+    #[serde(rename = "none")]
+    None,
+    #[serde(rename = "previous")]
+    Previous,
+    #[serde(rename = "file")]
+    File,
+    #[serde(rename = "data")]
+    Data,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CheckPasswd {
+    #[serde(rename = "type")]
+    variant: CheckPasswdType,
+    filename: Option<String>,
+    // Skip this for now, a separate file is easier
+    // and anyways we want to switch to sysusers
+    // entries: Option<Map<>String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TreeComposeConfig {
```

Comment:
If this is a human format (as YAML suggests) and newly introduced, I'd suggest to version this with a "kind+value" [tag](https://serde.rs/enum-representations.html#adjacently-tagged) as it will make future breaking changes easier to handle.

Reply:
Ooh that's *beautiful*.  Indeed that'd be useful here.  That said the plan is to deprecate this specific functionality entirely in favor of [sysusers](https://github.com/projectatomic/rpm-ostree/issues/49).